packages/astro/src/vite-plugin-astro-server/index.ts

```diff
@@ -63,6 +64,15 @@ async function handle500Response(viteServer: vite.ViteDevServer, origin: string,
 	writeHtmlResponse(res, 500, transformedHtml);
 }
 
+function getCustom404Route(config: AstroConfig, manifest: ManifestData) {
+	const relPages = config.pages.href.replace(config.projectRoot.href, '');
+	return manifest.routes.find((r) => r.component === relPages + '404.astro');
+}
+
+function log404(logging: LogOptions, pathname: string) {
+	info(logging, 'astro', msg.req({ url: pathname, statusCode: 404 }));
+}
```

Comment:
Not sure if this needs to be extracted just yet. One copy + paste doesn't bother me.

Reply:
This is called 3 times actually so I felt it was worth the abstraction. Your call though

Reply:
Fine with me!